### PR TITLE
Lazy render

### PR DESCRIPTION
**Pull request by semeleven**
Like in react-modal (https://github.com/reactjs/react-modal/blob/master/src/components/ModalPortal.js#L335) added the flag to render content inside tooltip after a show. It is needed, when content some hard - in my case datetimepicker. I think this is not needed for default, because of tooltips commonly light components.

### CODE REVIEW

**Comment by atomiks** [comment]
I'd like to keep the main wrapper as light as possible, this seems to introduce quite a bit of complexity.

Can't this functionality can be achieved in userland relatively easily:

```jsx
<Tippy 
  content={this.state.didShow ? <Datepicker /> : 'Loading...'} 
  onShow={() => this.setState({ didShow: true })}
>
  <button />
</Tippy>
```

That said, I don't mind exporting additional wrapper components that can be tree-shaken away by bundlers, such as:

- Multiple tippys on a single reference, `<MultipleTippy />`
- Lazy rendered content, `<LazyTippy />`
- Groups for delay, `<GroupTippy />`

**Comment by semeleven** [comment]
@atomiks No problem will try a separate component. Make the change

**Comment by atomiks** [comment]
Nice! 💙

**Comment by atomiks** [comment]
1. I was wondering because each time they show the tooltip it will need to be re-rendered right? Wouldn't it be better to keep it rendered if the expensive work has already been done? (this is how `lazy` works for the popperInstance in the native library) 

Btw, when you say "hard" does that mean "expensive"?

2. Right yeah I totally missed that

**Comment by KubaJastrz** [comment]
Yep, I'd change default behavior to keep the `content` component rendered when hidden, as the `Lazy` in the name suggests. And then add `destroyContentOnHide` prop specific only to `LazyTippy`.

File: test/LazyTippy.test.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  fireEvent,
+  cleanup,
+  waitForElement,
+  wait
+} from 'react-testing-library'
+
+afterEach(cleanup)
+
+describe('<LazyTippy />', () => {
+  test('tooltip content is not rendered before show', () => {
```
Comment (KubaJastrz) [inline test/LazyTippy.test.js]: Test suites should contain `expect` check. In this case
```js
expect(() => {
  class ...
}).not.toThrow();

File: test/LazyTippy.test.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    expect(onShow).toBeCalledTimes(1)
+    expect(onShow).toBeCalledWith(tip)
+    await wait(() => [
+      expect(onHidden).toBeCalledTimes(1),
+      expect(onHidden).toBeCalledWith(tip)
+    ])
+    done()
+  })
+
+  test('tooltip content is rendered after trigger without callbacks', async done => {
```
Comment (KubaJastrz) [inline test/LazyTippy.test.js]: No `expect` here too. I'm not even sure if this test checks for correct behavior or not. If you're relying on the error throwing, please do as described in previous comment

**Comment by atomiks** [comment]
Also, for treeshaking to work, I think we need a `"sideEffects": false` key in package.json.

**Comment by atomiks** [comment]
V2 structure is too different for this to be migrated. Feel free to make a new PR with this functionality

**Comment by itayganor** [comment]
This is huge and super important!
I was about to submit a new issue and already made [a codesandbox demo](https://codesandbox.io/s/tippyjs-react-always-mounts-5b3je) to demonstrate the non-lazyness (initially felt like a bug)

**Comment by atomiks** [comment]
@itayganor feel free to make a PR to make lazy default if possible without a breaking change. Or highlight this behavior on the docs so it's less expected maybe?

**Comment by itayganor** [comment]
@atomiks I'm trying to re-implement LazyTippy now to see if I could make it a PR.
But I found out that with singletons, `onShow` is also inherited from the singleton's source.
That means I can't rely on `onShow` in `LazyTippy`; The only way to work around this is to mention `'onShow'` in the `overrides` array of the singleton.

So these are the options:
1. Somehow check in `LazyTippy` if the given singleton has `onShow` in its `overrides` array. If not:
a. `console.error`
b. add it programmatically *(is it possible?)*
2. Add a note in the documentation about singletons and laziness. Make it clear `onShow` must be explicitly added to `overrides`.

**Comment by itayganor** [comment]
One more thing I noticed: I couldn't find any event related to singleton content changes. 
When the mouse is hovering a tippy (singleton target) and moves to another tippy (singleton target) `onShow` won't be called for the second.

**Comment by atomiks** [comment]
For your second comment, there are two update callbacks: `onBeforeUpdate` and `onAfterUpdate` which are called with content changes.

For your first comment, I'll need some time to think it through lol

**Comment by itayganor** [comment]
@atomiks Thanks for your replies.
Is there a way to determine if the current instance is gone hidden?
I tried using `onBeforeUpdate` and `onAfterUpdate`. I investigated what changes in `instance` or in `partialProps` but I couldn't find a good way to say if the current Tippy should get unrendered.

**Comment by atomiks** [comment]
You can find the full list of lifecycle callbacks here (starting with this): https://atomiks.github.io/tippyjs/v6/all-props/#onafterupdate

`onHidden` is the one you want I think.

**Comment by atomiks** [comment]
@itayganor I worked something out 🙌 . It requires a feature release in core but it should be fixed by tonight. You'll be able to create your own `<LazyTippy />` for now that works with singletons too, but it may be default in the future.

**Comment by atomiks** [comment]
This still renders "hidden" content but at least they only get mounted once the singleton is mounted, so it's an improvement. There's likely further optimizations you can perform.

```jsx
function Logger() {
  useEffect(() => {
    const interval = setInterval(() => console.log('is mounted'), 500);
    return () => clearInterval(interval);
  }, []);

  return 'logger';
}

function App() {
  const [source, target] = useSingleton();
  const [mounted, setMounted] = useState(false);

  return (
    <>
      <Tippy
        onShow={() => setMounted(true)}
        onHidden={() => setMounted(false)}
        singleton={source}
      />

      <Tippy content={mounted ? <Logger /> : ''} singleton={target}>
        <button>Hello</button>
      </Tippy>
      <Tippy content={mounted ? <Logger /> : ''} singleton={target}>
        <button>Hello</button>
      </Tippy>
    </>
  );
}
```

And a generic solution for other tippies: 

```jsx
function LazyTippy(props) {
  const [mounted, setMounted] = useState(false);

  const lazyPlugin = {
    fn: () => ({
      onShow: () => setMounted(true),
      onHidden: () => setMounted(false),
    }),
  };

  const computedProps = {...props};

  computedProps.plugins = [lazyPlugin, ...(props.plugins || [])];

  if (props.render) {
    computedProps.render = (...args) => (mounted ? props.render(...args) : '');
  } else {
    computedProps.content = mounted ? props.content : '';
  }

  return <Tippy {...computedProps} />;
}
```

**Pull request closed without merging**